### ISSUE TITLE
Merging PyTrees with `None` leaves

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#7828]
Let
```python
def split(pytree, masks):
    left  = jax.tree_map(lambda p, m:    p if m else None, pytree, masks)
    right = jax.tree_map(lambda p, m: None if m else p   , pytree, masks)
    return left, right
```
acting on `pytree`, `masks` which are PyTrees of identical structure; the latter assumed to have Boolean leaves.

I would like to define a function `merge` for which `merge(split(...))` is the identity.

(More generally, we might consider taking some tuple of PyTrees with common structure, and merging them all together, replacing the `None`s in each PyTree with the leaves (or subtrees) of the others.)

As far as I can see, there is no reasonable way to define such a `merge` at the moment, so I'm opening this issue in search of a solution. (Which I suspect would involve exposing additional functionality through JAX, so I've tagged this as a feature request.)

---

**Appendix:**

My current solution is to take inspiration from `vmap` and treat `None`s as PyTree leaves:

```python
_sentinel = object()

def _merge(*args):
    for arg in args:
        if arg is not _sentinel:
            return arg
    return _sentinel

def merge(*pytrees: PyTree):
    pytrees = [jax._src.tree_util._replace_nones(_sentinel, pytree) for pytree in pytrees]
    return jax.tree_map(_merge, *pytrees)
```

However of course this relies on private functionality. Possible solutions might be to:
- Expose `jax._src.tree_util._replace_nones` as public functionality;
  - I note that its implementation also seems to rely on private functionality, so it's not possible to duplicate as an end user.
- Or add an optional argument to `tree_map` that treats `None`s as leaves.
  - c.f. #2348

**Comment by cgarciae** [comment]
Hey @patrick-kidger ! 

If you set `is_leaf=lambda x: x is None` in `tree_map` and friends you can force `None`s as leafs. I used something similar in Treex for what you described.

**Comment by patrick-kidger** [comment]
Aha, that's a nice trick -- solves the problem perfectly. Thanks a lot!
